### PR TITLE
Aria2 module settings

### PR DESCRIPTION
**Pull request by timhae**
## Description of changes

aria2 module settings, added nixos test and myself as maintainer

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by timhae** [comment]
Result of `nixpkgs-review pr 303429` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

File: nixos/modules/services/networking/aria2.nix [hunk 1]
```diff
@@ -11,86 +9,122 @@ let
...
+      enable-rpc=true
+      listen-port=${lib.concatStringsSep "," (rangesToStringList cfg.listenPortRange)}
+      rpc-listen-port=${toString cfg.rpcListenPort}
+      save-session=${sessionFile}
+    '' + builtins.concatStringsSep "\n" (lib.mapAttrsToList
+      (key: value:
+        if builtins.typeOf value == "bool"
+        then "${key}=${lib.boolToString value}"
+        else "${key}=${toString value}")
+      cfg.settings) + "\n"
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:30]: Can this use `generators.toINI` (of some flavour)?

File: nixos/modules/services/networking/aria2.nix [hunk 2]
```diff
@@ -11,86 +9,122 @@ let
...
-          one of many WebUI (http://localhost:6800/ by default).
-
-          Targets are downloaded to ${downloadDir} by default and are
-          accessible to users in the "aria2" group.
-        '';
-      };
-      openPorts = mkOption {
-        type = types.bool;
+      enable = lib.mkEnableOption (lib.mdDoc ''
+        Whether or not to enable the headless Aria2 daemon service.
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:41]: `mkEnableOption` already add `Whether or not to enable` and a final `.`.

File: nixos/modules/services/networking/aria2.nix [hunk 3]
```diff
@@ -11,86 +9,122 @@ let
...
+          this module: `dir`, `listen-port`, `rpc-listen-port`, `enable-rpc`,
+          `save-session` and `conf-path`
+
+          [0]: https://aria2.github.io/manual/en/html/aria2c.html#synopsis
         '';
+        default = { };
+        type = lib.types.submodule {
+          freeformType = with lib.types; attrsOf (oneOf [ singleLineStr int float bool ]);
+        };
       };
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:98]: I take it `settings` is supposed to replace `extraArguments`.

If so, please add a `mkRemoveOptionModule` to inform users of the migration step.

File: nixos/modules/services/networking/aria2.nix [hunk 4]
```diff
@@ -11,86 +9,122 @@ let
...
     };
   };
 
-  config = mkIf cfg.enable {
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = !(cfg.settings ? dir);
+        message = "Use `services.aria2.downloadDir` to set the download directory.";
+      }
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:107]: Can we `mkRenamedOptionModule` to remove `downloadDir` and use `settings.dir` instead?

File: nixos/modules/services/networking/aria2.nix [hunk 5]
```diff
@@ -11,86 +9,122 @@ let
...
+        message = "RPC has to be enabled, the module takes care of that.";
+      }
+      {
+        assertion = !(cfg.settings ? conf-path);
+        message = "Can not change the configuration file path.";
+      }
+      {
+        assertion = !(cfg.settings ? listen-port);
+        message = "Use `services.aria2.listenPortRange` to set the listen port range.";
+      }
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:119]: `mkRenamedOptionModule`?

File: nixos/modules/services/networking/aria2.nix [hunk 6]
```diff
@@ -11,86 +9,122 @@ let
...
+      {
+        assertion = !(cfg.settings ? rpc-listen-port);
+        message = "Use `services.aria2.rpcListenPort` to set the rpc listen port.";
+      }
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:123]: `mkRenamedOptionModule`?

File: nixos/modules/services/networking/aria2.nix [hunk 7]
```diff
@@ -120,17 +154,20 @@ in
           touch "${sessionFile}"
         fi
         cp -f "${settingsFile}" "${settingsDir}/aria2.conf"
+        chmod +w "${settingsDir}/aria2.conf"
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:157]: Why?

File: nixos/tests/aria2.nix [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import ./make-test-python.nix ({ pkgs, ... }:
+let
+  rpcSecret = "supersecret";
+  curlBody = {
+    jsonrpc = 2.0;
+    id = 1;
+    method = "aria2.getVersion";
+    params = [ "token:${rpcSecret}" ];
+  };
+in
```
Comment (ambroisie) [inline nixos/tests/aria2.nix:10]:
```suggestion
  rpcListenPort = 6800;
in
```

File: nixos/tests/aria2.nix [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+rec {
+  name = "aria2";
+
+  nodes.machine = {
+    environment.etc."aria2Rpc".text = rpcSecret;
+    services.aria2 = {
+      enable = true;
+      rpcSecretFile = "/etc/aria2Rpc";
+      rpcListenPort = 6800;
```
Comment (ambroisie) [inline nixos/tests/aria2.nix:19]:
```suggestion
      inherit rpcListenPort;
```

File: nixos/tests/aria2.nix [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        console-log-level = "warn";
+      };
+    };
+  };
+
+  testScript = ''
+    machine.start()
+    machine.wait_for_unit("aria2.service")
+    curl_cmd = 'curl --fail-with-body -X POST -H "Content-Type: application/json" \
+                -d \'${builtins.toJSON curlBody}\' http://localhost:${toString nodes.machine.services.aria2.rpcListenPort}/jsonrpc'
```
Comment (ambroisie) [inline nixos/tests/aria2.nix:35]:
```suggestion
                -d \'${builtins.toJSON curlBody}\' http://localhost:${toString rpcListenPort}/jsonrpc'
```

**Review by ambroisie — CHANGES_REQUESTED** [review]
I would split the addition of the test from the new `settings` option.

Could you also either not change the formatting, or split it into a different commit?

Otherwise thank you for adding a test to the module! This makes me much more confident for refactoring it now :-)

File: nixos/modules/services/networking/aria2.nix (hunk 7 above)
Comment (timhae) [reply nixos/modules/services/networking/aria2.nix:157]: To be honest, I don't know why it worked before. Files in the nix store are not writable and copying the file from the nix store to some other directory doesn't change that. See the `set -x` test output here:
```
machine # [    8.628432] aria2-pre-start[794]: + cp -f /nix/store/9pysl0bgc35wagb4qyrni43a0sgmjfmm-aria2.conf /var/lib/aria2/aria2.conf
machine # [    8.631637] systemd[1]: Finished Permit User Sessions.
machine # [    8.644510] systemd[1]: Started Getty on tty1.
machine # [    8.648132] systemd[1]: Reached target Login Prompts.
machine # [    8.650737] aria2-pre-start[794]: + ls -la /var/lib/aria2
machine # [    8.697284] aria2-pre-start[801]: total 16
machine # [    8.698995] aria2-pre-start[801]: drwxrwx--- 3 aria2 aria2 4096 Apr 16 18:57 .
machine # [    8.702515] aria2-pre-start[801]: drwxr-xr-x 8 root  root  4096 Apr 16 18:57 ..
machine # [    8.704837] aria2-pre-start[801]: -r--r--r-- 1 aria2 aria2  307 Apr 16 18:57 aria2.conf
machine # [    8.706943] aria2-pre-start[801]: -rw-r--r-- 1 aria2 aria2    0 Apr 16 18:57 aria2.session
machine # [    8.709370] aria2-pre-start[801]: drwxrwx--- 2 aria2 aria2 4096 Apr 16 18:57 Downloads
machine # [    8.712158] aria2-pre-start[794]: + chmod +w /var/lib/aria2/aria2.conf
machine # [    8.715914] aria2-pre-start[794]: + ls -la /var/lib/aria2
machine # [    8.731449] aria2-pre-start[803]: total 16
machine # [    8.732721] aria2-pre-start[803]: drwxrwx--- 3 aria2 aria2 4096 Apr 16 18:57 .
machine # [    8.734577] aria2-pre-start[803]: drwxr-xr-x 8 root  root  4096 Apr 16 18:57 ..
machine # [    8.736273] aria2-pre-start[803]: -rw-r--r-- 1 aria2 aria2  307 Apr 16 18:57 aria2.conf
machine # [    8.738103] aria2-pre-start[803]: -rw-r--r-- 1 aria2 aria2    0 Apr 16 18:57 aria2.session
machine # [    8.739719] aria2-pre-start[803]: drwxrwx--- 2 aria2 aria2 4096 Apr 16 18:57 Downloads
```
So I need to chmod in order to add the rpc secret.

**Comment by timhae** [comment]
Thanks for your comments, I have tried to address all of them. Please let me know what you think.

File: nixos/modules/services/networking/aria2.nix [hunk 11]
```diff
@@ -1,98 +1,137 @@
...
+  };
 in
 {
   imports = [
-    (mkRemovedOptionModule [ "services" "aria2" "rpcSecret" ] "Use services.aria2.rpcSecretFile instead")
+    (lib.mkRemovedOptionModule [ "services" "aria2" "rpcSecret" ] "Use services.aria2.rpcSecretFile instead")
+    (lib.mkRemovedOptionModule [ "services" "aria2" "extraArguments" ] "Use services.aria2.settings instead")
+    (lib.mkRemovedOptionModule [ "services" "aria2" "downloadDir" ] "Use services.aria2.settings.dir instead")
+    (lib.mkRemovedOptionModule [ "services" "aria2" "listenPortRange" ] "Use services.aria2.settings.listen-port instead")
+    (lib.mkRemovedOptionModule [ "services" "aria2" "rpcListenPort" ] "Use services.aria2.settings.rpc-listen-port instead")
```
Comment (ambroisie) [inline nixos/modules/services/networking/aria2.nix:39]: Unless I'm mistaken, we can use `mkRenamedOptionModule` for these, can't we?

```suggestion
    (lib.mkRenamedOptionModule [ "services" "aria2" "downloadDir" ] [ "services" "aria2" "settings" "dir" ] )
    (lib.mkRenamedOptionModule [ "services" "aria2" "listenPortRange" ] [ "services" "aria2" "settings" "listen-port" ])
    (lib.mkRenamedOptionModule [ "services" "aria2" "rpcListenPort" ] [ "services" "aria2" "settings" "rpc-listen-port" ])
```

Comment (timhae) [reply nixos/modules/services/networking/aria2.nix]: agreed

**Review by ambroisie — APPROVED** [review]

**Comment by jys1670** [comment]
I am getting `error: The option 'services.aria2.settings' is used but not defined.` on the latest nixos-unstable.
Was this meant to happen? From a brief look, it seems that `services.aria2.settings` should be defined with some default values.
@timhae

**Comment by timhae** [comment]
@jys1670 can you please post your `services.aria2` settings in your nixconfig?

**Comment by jys1670** [comment]
> @jys1670 can you please post your `services.aria2` settings in your nixconfig?
Yeah, sure:
```Nix
  services.aria2 = {
    enable = true;
    rpcSecretFile = "${pkgs.writeText "aria" "aria2rpc\n"}";
  };
```
It's working fine if I revert this PR. Should something else be defined for default settings on latest nixos-unstable?

**Comment by timhae** [comment]
Sorry for breaking your config, can you change that to:
```
  services.aria2 = {
    enable = true;
    rpcSecretFile = "${pkgs.writeText "aria" "aria2rpc\n"}";
    settings = { };
  };
```

**Comment by jys1670** [comment]
> Sorry for breaking your config, can you change that to:
> 
> ```
>   services.aria2 = {
>     enable = true;
>     rpcSecretFile = "${pkgs.writeText "aria" "aria2rpc\n"}";
>     settings = { };
>   };
> ```

Thanks! This fixed the issue for me.

**Comment by timhae** [comment]
created PR #320035

**Pull request merged**